Fill out XML spec document (#247)
rojo-rbx/rbx-dom: docs/xml.md

	- [ColorSequence](#colorsequence)
	- [Content](#content)
	- [CoordinateFrame](#coordinateframe)
	- [double](#double) (Float64)
	- [Faces](#faces)
	- [float](#float) (Float32)
	- [Font](#font)
	- [int](#int) (Int32)
	- [int64](#int64)
	- [NumberRange](#numberrange)
	- [NumberSequence](#numbersequence)
	- [Optional](#optional)
	- [PhysicalProperties](#physicalproperties)
	- [ProtectedString](#protectedstring)
	- [Ray](#ray)
	- [Rect2D](#rect2d)
	- [Ref](#ref) (Referent)
	- [SharedString][SharedString-use] (property type)
	- [string](#string)
	- [token](#token) (Enum)
	- [UDim](#udim)
	- [UDim2](#udim2)
	- [UniqueId](#uniqueid)
	- [Vector2](#vector2)
	- [Vector3](#vector3)
	- [Vector3int16](#vector3int16)

## File Structure
Roblox XML files consist of a single `<roblox>` element, which contain a sequence of other elements. The basic layout of the file structure is as follows:

- Exactly one `<roblox>` element
	- Zero or more `<Meta>` elements
	- Zero or more `<External>` elements
	- One or more `<Item>` elements
	- One `<Properties>` element
		- Zero or more type elements
	- Zero or more `<Item>` elements (this may nest infinitely)
	- Zero or one `<SharedStrings>` elements

> 		- Zero or more `<SharedString>` elements

Author: Tried to improve the entire section. Eager to hear your opinion.

> **To be accepted by Roblox, files MUST start with `<roblox` and end with `</roblox>`, with no whitespace.**

Author: My intention was to indicate what Roblox does and doesn't accept (several webpoints reject any file that doesn't begin with `<roblox` and end with `</roblox>`, which is why the binary format begins and ends with those) but in hindsight it's confusing and out of place so I've simply removed it.

> 

Author: The intent is to differ between suggestions and hard limitations; certain things will get a file outright rejected by Roblox or cause it to misinterpret data, and it's important to note what those are with emphasis because a lot of things are just "best practices" rather than hard limitations.

I'm open to changing this, but the current capitalization follows the IETF's recommendations in [RFC-2119](https://www.rfc-editor.org/rfc/rfc2119).

> 

Reviewer: Yes, it makes sense. There's perhaps an argument to be had about how formal these documents ought to be, but I think I'm fine with it as long as we say that we're following RFC 2119, link to it, and carefully consider each usage (particularly of SHOULD). We might also think about using the same convention in the binary spec to stay consistent.

> 

Author: I'll do a pass on this to re-evaluate where we should use these in a bit. I meant to do it before finalizing the PR but evidently forgot. I know I specifically overused `SHOULD` but as I'm reading it back I misused `MAY` as well.

Wrt the binary spec, I agree and I'm planning to do a cleanup pass on it Soon:tm: because I think we can do some of it better. When I do that, I'll see where they can fit, but with binary files it starts to get weird since most of it is pretty much set in stone.

Given how we are to my knowledge the de-facto repository for knowledge on these file formats outside of Roblox (I only know of one other complete spec file for both formats), I think we should try to be as formal as possible about documenting them. People will take us at our word and I'd like to not disappoint them.

## roblox

This element is the root of the file. There MUST be one `roblox` element in the file.

The following attributes are required for this element:

| Name      | Contents                                                                         |
|:----------|:---------------------------------------------------------------------------------|
| `version` | The version of the format this document contains. This MUST be `4` at this time. |

> All other attributes are ignored by Roblox, including those defined by Roblox Studio when exporting files.

Author: Given they're not necessary for the file, I don't think including them in the spec file is a good idea. I would like this to only include information that's relevant for people reading or writing a file.

That said, I am absolutely open to writing the attributes studio generates down *somewhere*. Given there's several elements and an attribute I'm already omitting (they're used internally to save the IDE state), it might make sense to have a supplementary file for data that's useful but not important like this.

A `token` representing `Enum.NormalId.Left` would appear as follows:

```xml
<token name="TokenExample">3</token>
```

### UDim

The `UDim` data type is represented as a sequence of two child elements indicating the `Scale` and `Offset` components of the value. These child elements are named `S` and `O`. The `S` element is a [`float`](#float) and the `O` element is a [`int`](#int).

A `UDim` with the value `{0.15625, 1337}` would appear as follows:

```xml
<UDim name="UDimExample">
	<S>0.15625</S>
	<O>1337</O>
</UDim>
```

### UDim2

The `UDim2` data type is represented as a sequence of four child elements indicating the `X.Scale`, `X.Offset`, `Y.Scale`, `Y.Offset` components of the value. These elements are named `XS`, `XO`, `YS`, and `YO`. The `XS` and `YS` elements are [`float`](#float) values and the `XO` and `YO` elements are [`int`](#int) values.

A `UDim2` with the value `{0.15625, 1337}, {-123, 456}` would appear as follows:

```xml
<UDim2 name="UDim2Example">
	<XS>0.15625</XS>
	<XO>1337</XO>
	<YS>-123</YS>
	<YO>456</YO>
</UDim2>
```

### UniqueId

> The `UniqueId` data type is represented as hexadecimal-encoded sequence of `16` bytes. These bytes may be split in three distinct groups, representing components of the `UniqueId`:

Reviewer: ```suggestion
The `UniqueId` data type is represented as hexadecimal-encoded sequence of `16` bytes. These bytes are split into three distinct groups, representing components of the `UniqueId`:
```

There are no attributes required for this element.

Every child of this element is a [Type Element](#type-elements) and represents exactly one property of an `Instance`.

## SharedStrings

This element acts as a repository for [`SharedString` definitions][SharedString-def]. There MAY be zero or one `SharedStrings` elements per file. When present, the `SharedStrings` element MUST be under the `roblox` element.

There are no attributes required for this element.

## SharedString
[SharedString-def]: #sharedstring

This element defines a single `SharedString` value for reference by [Type Elements](#type-elements). There MAY be zero or more `SharedString` elements per file. `SharedString` elements MUST be under the `SharedStrings` element.

This element shares a name with a type element. That element is documented [here][SharedString-use].

The following attributes are REQUIRED for this element:

| Name  | Contents                                                               |
|:------|:-----------------------------------------------------------------------|
| `md5` | A unique identifier for this element, for reference by a type element. |

> The value of `md5` MUST be unique across all `SharedString` definitions. Despite the name, the value does not have to be the MD5 hash of the `SharedString` contents.

Reviewer 2: They don't have to be the MD5 hash of the SharedString contents, but does it need to be MD5 at all? (I suspect no?)

Author: The attribute has be named `md5`, if that's what you mean.

The content of `SharedString` elements MUST be Base64 encoded.

> <!-- TODO: Verify what form of Base64 and put it here-->

Reviewer 2: :eyes:

### BrickColor

The `BrickColor` data type is represented by a single 32-bit integer that represents the `Number` of the value.

Roblox encodes this type with the element name `int` but also accepts `BrickColor`. It is preferred that `BrickColor` be used.

A `BrickColor` with the value `Medium Stone Grey` (whose number is `194`) should appear as follows:

```xml
<BrickColor name="BrickColorExample">194</BrickColor>
```

### Color3

The `Color3` data type is represented by three child elements named `R`, `G`, and `B`. These elements contain the value of that component as written as a [`float`](#float).

A `Color3` with the value `INF, 1337, 0.15625` appears as follows:

```xml
<Color3 name="Color3Example">
	<R>INF</R>
	<G>1337</G>
	<B>0.15625</B>
</Color3>
```

### Color3uint8

The `Color3uint8` data type is represented by a single unsigned 32-bit integer that is the `R`, `G`, and `B` components of the color (as integers in the range 0 to 255) packed into the lower 24 bits of the number. This integer is little-endian and written in the order `G`, `B`, `R`.

The upper 8 bits of the value SHOULD be filled with `FF` (in hexadecimal). This is to maintain compatibility with Roblox.

> A `Color3uint8` with the value `96, 64, 32` appears as follows:

Reviewer 2: Might be nice to show the steps in getting to `4284497952`, like the above

Author: I've done my best to specify but I'm not sure how best to explain it while keeping tonally consistent. On the plus side, I discovered an issue with the info I'd written while explaining, so that's fixed now.
